action/protocol/staking/ethabi/common/bucketsbycandidate.go

```diff
@@ -0,0 +1,76 @@
+package common
+
+import (
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/iotexproject/iotex-proto/golang/iotexapi"
+	"github.com/iotexproject/iotex-proto/golang/iotextypes"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/iotexproject/iotex-core/action/protocol"
+)
+
+// BucketsByCandidateStateContext context for BucketsByCandidate
+type BucketsByCandidateStateContext struct {
+	*protocol.BaseStateContext
+}
+
+func NewBucketsByCandidateStateContext(data []byte, methodABI *abi.Method, apiMethod iotexapi.ReadStakingDataMethod_Name) (*BucketsByCandidateStateContext, error) {
+	paramsMap := map[string]interface{}{}
+	ok := false
+	if err := methodABI.Inputs.UnpackIntoMap(paramsMap, data); err != nil {
+		return nil, err
+	}
+	var candName string
+	if candName, ok = paramsMap["candName"].(string); !ok {
+		return nil, ErrDecodeFailure
+	}
+	var offset, limit uint32
+	if offset, ok = paramsMap["offset"].(uint32); !ok {
+		return nil, ErrDecodeFailure
+	}
+	if limit, ok = paramsMap["limit"].(uint32); !ok {
+		return nil, ErrDecodeFailure
+	}
+
+	method := &iotexapi.ReadStakingDataMethod{
+		Method: apiMethod,
+	}
+	methodBytes, err := proto.Marshal(method)
+	if err != nil {
+		return nil, err
+	}
+	arguments := &iotexapi.ReadStakingDataRequest{
+		Request: &iotexapi.ReadStakingDataRequest_BucketsByCandidate{
+			BucketsByCandidate: &iotexapi.ReadStakingDataRequest_VoteBucketsByCandidate{
+				CandName: candName,
+				Pagination: &iotexapi.PaginationParam{
+					Offset: offset,
+					Limit:  limit,
+				},
+			},
+		},
+	}
+	argumentsBytes, err := proto.Marshal(arguments)
+	if err != nil {
+		return nil, err
+	}
+	return &BucketsByCandidateStateContext{
+		&protocol.BaseStateContext{
+			Parameter: &protocol.Parameters{
+				MethodName: methodBytes,
+				Arguments:  [][]byte{argumentsBytes},
+			},
+			Method: methodABI,
+		},
+	}, nil
+}
+
+// EncodeToEth encode proto to eth
+func (r *BucketsByCandidateStateContext) EncodeToEth(resp *iotexapi.ReadStateResponse) (string, error) {
+	var result iotextypes.VoteBucketList
+	if err := proto.Unmarshal(resp.Data, &result); err != nil {
+		return "", err
+	}
+
+	return EncodeVoteBucketListToEth(r.Method.Outputs, &result)
```

Comment:
looks like this code is the same for all the new contexts you added. Can move to `protocol.BaseStateContext` and become a common method?

Reply:
It's acceptable to define `EncodeVoteBucketListToEth` same pkg with  `BucketEth`, and moving will cause cycle dependency problem